### ISSUE TITLE
Implement a different mimetype guess() approach

### ISSUE THREAD

**Issue by DiegoPino** [esmero/strawberryfield#153]
# What is this?

I may need some help with PROS and CONS here but my tests so far are resulting in success

Right now the ::guess function does this:

```PHP
/**
   * {@inheritdoc}
   */
  public function guess($path) {
    if ($this->mapping === NULL) {
      $mapping = $this->defaultMapping;
      // Allow modules to alter the default mapping.
      $this->moduleHandler->alter('file_mimetype_mapping', $mapping);
      $this->mapping = $mapping;
    }

    $extension = '';
    $file_parts = explode('.', \Drupal::service('file_system')->basename($path));

    // Remove the first part: a full filename should not match an extension.
    array_shift($file_parts);

    // Iterate over the file parts, trying to find a match.
    // For my.awesome.image.jpeg, we try:
    //   - jpeg
    //   - image.jpeg, and
    //   - awesome.image.jpeg
    while ($additional_part = array_pop($file_parts)) {
      $extension = strtolower($additional_part . ($extension ? '.' . $extension : ''));
      if (isset($this->mapping['extensions'][$extension])) {
        return $this->mapping['mimetypes'][$this->mapping['extensions'][$extension]];
      }
    }

    return 'application/octet-stream';
  }
````

You see that in this case the attempts are made from more general to more precise.
This leaves our warc.gz, and tar.gz out of the selection defaulting to application/gzip even if have provide our own mappings
In most cases extensions are mapped to a single mime but in Digital repositories people are quite aware of that and also we want to use precise mime detection to be able to decide if an asset qualifies or not be used by a certain formatter

My implementation (using our own existing service that provides also a reverse mimetype to extension) does this

```PHP
/**
   * {@inheritdoc}
   */
  public function guess($path) {
    if ($this->mapping === NULL) {
      $mapping = $this->defaultMapping;
      // Allow modules to alter the default mapping.
      $this->moduleHandler->alter('file_mimetype_mapping', $mapping);
      $this->mapping = $mapping;
    }

    $extension = '';
    $file_parts = explode('.', \Drupal::service('file_system')->basename($path));

    // Remove the first part: a full filename should not match an extension.
    array_shift($file_parts);

    // Iterate over the file parts, trying to find all matches.
    // For my.awesome.image.jpeg, we try and accumulate:
    //   - jpeg
    //   - image.jpeg, and
    //   - awesome.image.jpeg
    $qualified = ['application/octet-stream'];
    while ($additional_part = array_pop($file_parts)) {
      $extension = strtolower($additional_part . ($extension ? '.' . $extension : ''));
      if (isset($this->mapping['extensions'][$extension])) {
        $qualified[] = $this->mapping['mimetypes'][$this->mapping['extensions'][$extension]];
      }
    }

    // We return only the last one giving any more precise one to have a chance.
    return end($qualified);

  }
```
Basically I accumulate all from more general to more precise and fetch the last result and return it.In most cases it will be the same but in cases of multiple matches based on `.` this actually works better

@pcambra @giancarlobi @patdunlavey @alliomeria any concerns? I will make a pull but if no concerns i may go for this in RC2

**Comment by DiegoPino** [comment]
Solved via https://github.com/esmero/strawberryfield/commit/1ee0cff7cbd6bae64530da8dc3421a4292782e9e
